### PR TITLE
Added uri builder to construct query params

### PR DESCRIPTION
**Pull request by mmoayyed**
Handles #104 

I adapted the HttpClient URIBuilder so I can add/remove query parameters easily. Added 4-5 test cases to make sure service url construction is handled without and with encoding.

### CODE REVIEW

File: cas-client-core/src/main/java/org/jasig/cas/client/util/URIBuilder.java [hunk 1]
```diff
@@ -0,0 +1,604 @@
...
+import java.util.Iterator;
+import java.util.List;
+import java.util.regex.Pattern;
+
+/**
+ * A utility class borrowed from apache http-client to build uris.
+ * @author Misagh Moayyed
+ * @since 3.4
+ */
+public final class URIBuilder {
```
Comment (leleuj) [inline cas-client-core/src/main/java/org/jasig/cas/client/util/URIBuilder.java]: It's a copy / paste from the HTTP client library, isn't it?

Comment (mmoayyed) [inline cas-client-core/src/main/java/org/jasig/cas/client/util/URIBuilder.java]: 80% yes.

Comment (leleuj) [inline cas-client-core/src/main/java/org/jasig/cas/client/util/URIBuilder.java]: Too bad we don't have such a class in a `common-stuff` library...

**Comment by leleuj** [comment]
+1

File: cas-client-core/src/main/java/org/jasig/cas/client/util/URIBuilder.java (hunk 1 above)
Comment (battags) [inline cas-client-core/src/main/java/org/jasig/cas/client/util/URIBuilder.java]: Is our license header the correct header then?

File: cas-client-core/src/main/java/org/jasig/cas/client/util/CommonUtils.java [hunk 2]
```diff
@@ -259,70 +260,59 @@ private static boolean requestIsOnStandardPort(final HttpServletRequest request)
...
+     * @param service the configured service url (this will be used if not null)
+     * @param serverNames the server name to  use to constuct the service url if the service param is empty.  Note, prior to CAS Client 3.3, this was a single value.
+     *           As of 3.3, it can be a space-separated value.  We keep it as a single value, but will convert it to an array internally to get the matching value. This keeps backward compatability with anything using this public
+     *           method.
+     * @param serviceParameterName the service parameter name to remove (i.e. service)
+     * @param artifactParameterName the artifact parameter name to remove (i.e. ticket)
+     * @param encode whether to encode the url or not (i.e. Jsession).
+     * @return the service url to use.
+     */
     public static String constructServiceUrl(final HttpServletRequest request, final HttpServletResponse response,
```
Comment (battags) [inline cas-client-core/src/main/java/org/jasig/cas/client/util/CommonUtils.java]: we need to confirm this change won't break Spring Security integration.  Some of our CommonUtils are "public API" :-(

File: cas-client-core/src/main/java/org/jasig/cas/client/util/URIBuilder.java [hunk 3]
```diff
@@ -0,0 +1,604 @@
...
+    private boolean encode;
+    private String fragment;
+    private String encodedFragment;
+
+    /**
+     * Constructs an empty instance.
+     */
+    public URIBuilder() {
+        super();
+        this.port = -1;
```
Comment (battags) [inline cas-client-core/src/main/java/org/jasig/cas/client/util/URIBuilder.java]: -1?

File: cas-client-core/src/main/java/org/jasig/cas/client/util/URIBuilder.java (hunk 1 above)
Comment (mmoayyed) [inline cas-client-core/src/main/java/org/jasig/cas/client/util/URIBuilder.java]: Their license is strictly Apache v2, while ours has a bit of Apereo in it. I could include them both, but I don't know how the license plugin will react.

File: cas-client-core/src/main/java/org/jasig/cas/client/util/CommonUtils.java (hunk 2 above)
Comment (mmoayyed) [inline cas-client-core/src/main/java/org/jasig/cas/client/util/CommonUtils.java]: @rwinch Is this change something that might break the integration? I could add an overloaded method here, but I dont think I can make it to work correctly.

File: cas-client-core/src/main/java/org/jasig/cas/client/util/URIBuilder.java (hunk 3 above)
Comment (mmoayyed) [inline cas-client-core/src/main/java/org/jasig/cas/client/util/URIBuilder.java]: See http://www.grepcode.com/file/repo1.maven.org/maven2/org.apache.httpcomponents/httpclient/4.4/org/apache/http/client/utils/URIBuilder.java#71

Comment (mmoayyed) [inline cas-client-core/src/main/java/org/jasig/cas/client/util/URIBuilder.java]: I guess to better answer your question; I am not sure why it's set to -1 initially. 0 might imply a default port for the scheme chosen. and there are other checks down below that specifically check for -1.

**Comment by mmoayyed** [comment]
Any other comments? Are we good to move fwd with this?

**Comment by battags** [comment]
+1

**Pull request merged**